lib/public/OCM/Events/BeforeResourceTypesGetEvent.php

```diff
@@ -0,0 +1,37 @@
+<?php
+
+declare(strict_types=1);
+/*
+ * @copyright Copyright (c) 2023 Joas Schilling <coding@schilljs.com>
+ *
+ * @author Joas Schilling <coding@schilljs.com>
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCP\OCM\Events;
+
+use OCP\EventDispatcher\Event;
+
+/**
+ * Use this event to register additional OCM resources before the API returns
+ * them in the OCM provider list and capability
+ *
+ * @since 28.0.0
+ */
+class BeforeResourceTypesGetEvent extends Event {
```

Comment:
```suggestion
class BeforeResourceTypesReturnedEvent extends Event {
```

or something else that makes the purpose of the event clearer?

Reply:
ResourceTypeRegisterEvent?

Reply:
and could we improve the DX by passing the IOCMProvider implementation into the event and exposing a getter?

Reply:
Well the main job is not getting but adding,

See https://github.com/nextcloud/spreed/pull/10692/files#diff-3b678f93168d9f4120e050fae0237b8f70b8a03c8a3a3d9c1f81e63f05ffa676R56 for a sample listener

Reply:
Ah you mean a getter to get the provider? Can also do that.

Reply:
> why do I create a new resource but then add a resource type? should it be createResourceType or addResource?

That's a good question but the inconsitancy is already part of the public api 🙈

We could add a `registerResourceType()` directly to the event, so consumers don't even need to bother with the `IProvider` and `IOCMResource` themselves? 🤔 sounds good?

Reply:
sounds good!

Reply:
> ooooor we register the resource (types) via the application class?

That sounds a bit bloating for the fact we have 2 federation things only (files+ talk soon)